aries_cloudagent/connections/models/diddoc_v2/diddoc.py

```diff
@@ -0,0 +1,266 @@
+"""
+DID Document classes.
+
+Copyright 2021 Sicpa
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+import logging
+
+from typing import Union
+
+from .publickey import PublicKey
+from .service import Service
+from .util import canon_did
+from .schemas.diddocschema import DIDDocSchema
+
+LOGGER = logging.getLogger(__name__)
+
+
+class DIDDoc:
```

Comment:
Similar to what is found in the `ResolvedDIDDoc` class, when given a DID URL that matches the ID of this DID Document, I want to be able to dereference the URL to the correct part of the document. For instance, given the following DID Document:
```json
{
   "@context": "https://w3c.org/whatever/the/url/is",
   "id":"did:example:1234abcd",
   "authentication":[
      {
         "controller":"LjgpST2rjsoxYegQDRm7EL",
         "id":"3",
         "publicKeyPem":"-----BEGIN PUBLIC X...",
         "type":"RsaVerificationKey2018",
         "usage":"signing"
      }
   ],
   "publicKey":[
      {
         "controller":"LjgpST2rjsoxYegQDRm7EL",
         "id":"3",
         "publicKeyPem":"-----BEGIN PUBLIC X...",
         "type":"RsaVerificationKey2018",
         "usage":"signing"
      }
   ],
   "service":[
      {
         "id":"1",
         "priority":1,
         "recipientKeys":[
            "~XXXXXXXXXXXXXXXX",
            "did:sov:LjgpST2rjsoxYegQDRm7EL#keys-1"
         ],
         "routingKeys":[
            "did:sov:LjgpST2rjsoxYegQDRm7EL#keys-4"
         ],
         "serviceEndpoint":"LjgpST2rjsoxYegQDRm7EL;2",
         "type":"one"
      }
   ]
}
```

Calling `did_doc.dereference("did:example:1234abcd#1")` should return a `Service` representing the service block. When dereferencing `did:example:1234abcd#3`, it should return a `PublicKey`/`VerificationMethod` instance representing the public key. In implementing this, we should take advantage of the `DID` and `DIDUrl` classes already in the resolver package that Adam and I have been working on. You can refer to the tests for the `ResolvedDIDDoc` class for more details on how this should operate.

Reply:
This might be thought of as similar to calling `document.getElementById("element-id")` on a web page in JavaScript.